contracts/lib.rs

```diff
@@ -1672,1661 +1527,1788 @@ pub mod prosopo {
+                        assert!(contract.check_not_admin(*acc).is_ok());
+                        set_caller(*acc);
+                        assert!(contract.check_caller_admin().is_err());
+                    }
                 }
             }
-            assert!(contract.providers.get(provider_account).is_none());
-            assert!(!contract
-                .provider_accounts
-                .get(ProviderState {
-                    status: GovernanceStatus::Deactivated,
-                    payee: Payee::Dapp
-                })
-                .unwrap()
-                .contains(&provider_account));
-        }
 
-        /// Test provider update with service_origin error
-        #[ink::test]
-        fn test_provider_update_with_service_origin_error() {
-            let operator_accounts = get_operator_accounts();
-            let mut contract = Prosopo::default(
-                operator_accounts,
-                STAKE_DEFAULT,
-                STAKE_DEFAULT,
-                10,
-                1000000,
-                0,
-                1000,
-            );
+            #[ink::test]
+            fn test_set_admin() {
+                // always set the caller to the unused account to start, avoid any mistakes with caller checks
+                set_caller(default_unused_account());
+                let accounts = default_accounts();
+
+                let mut contract = default_contract();
+                let old_admin = contract.admin;
+                let new_admin = accounts.admins[1];
+                assert_ne!(old_admin, new_admin);
+
+                contract.check_admin(old_admin).unwrap();
+                contract.check_not_admin(new_admin).unwrap();
+
+                set_caller(old_admin);
+                contract.set_admin(new_admin).unwrap();
+
+                contract.check_admin(new_admin).unwrap();
+                contract.check_not_admin(old_admin).unwrap();
+            }
+
+            #[ink::test]
+            fn test_set_admin_unauthorised() {
+                // always set the caller to the unused account to start, avoid any mistakes with caller checks
+                set_caller(default_unused_account());
+                let accounts = default_accounts();
+
+                let mut contract = default_contract();
+                let old_admin = contract.admin;
+                let new_admin = accounts.admins[1];
+                assert_ne!(old_admin, new_admin);
+
+                contract.check_admin(old_admin).unwrap();
+                contract.check_not_admin(new_admin).unwrap();
+
+                // can only call set_admin from the current admin account (old admin)
+                set_caller(new_admin);
+                contract.set_admin(new_admin).unwrap_err();
+            }
+
+            #[ink::test]
+            fn test_ctor_caller_admin() {
+                // always set the caller to the unused account to start, avoid any mistakes with caller checks
```

Comment:
What is the purpose of this unused account?

Reply:
The idea behind this is that I spotted cases where accounts were in collision, e.g. a user set to [u1; 32] and a provider the same, say. This _can_ invalidate the caller guards on various functions, so to prevent this I have made this concept of an "unused" account which we can always rely on to be used for nothing. Consequently, you can set the caller to the unused account to be certain that we're not calling a function from an account which has another purpose.

More broadly speaking, I have made a batch of accounts in the unit tests for different purposes, e.g. 10x accounts for users, 10x accounts for providers, etc. By changing our unit tests to use these accounts for each purpose (which I am yet to finish for all unit tests, only new ones so far) we can be more certain on avoiding collisions in duplicate account usages. Also, it allows for setup methods to be used based on these accounts to decrease the duplicate code amongst unit tests (again, in the works)

Reply:
I have now changed this so the accounts are in the following format:
[ TYPE_PREFIX, index as 16x bytes, zeros ]

The type prefix indicates their roles, e.g. admins begin with 1, providers 2, etc.

The index is the index of the account, so the first admin is [1, 1, 0, ... ], the second is [1, 2, 0, ... ], the third [1, 3, 0, ... ], etc. This allows accounts to be generated on the fly for different account roles